src/Components/Components/src/Routing/Router.cs

```diff
@@ -152,19 +153,32 @@ private static string StringUntilAny(string str, char[] chars)
 
         private void RefreshRouteTable()
         {
-            var assemblies = AdditionalAssemblies == null ? new[] { AppAssembly } : new[] { AppAssembly }.Concat(AdditionalAssemblies);
-            var assembliesSet = new HashSet<Assembly>(assemblies);
+            if (!PreferExactMatches)
+            {
+                RefreshRouteTableLegacy();
+                return;
+            }
+
+            var routeKey = new RouteKey(AppAssembly, AdditionalAssemblies);
 
-            if (!_assemblies.SetEquals(assembliesSet))
+            if (!routeKey.Equals(_currentRouteKey))
             {
-                Routes = PreferExactMatches
-                    ? RouteTableFactory.Create(assemblies)
-                    : LegacyRouteTableFactory.Create(assemblies);
-                _assemblies.Clear();
-                _assemblies.UnionWith(assembliesSet);
+                _currentRouteKey = routeKey;
+                Routes = RouteTableFactory.Create(routeKey);
             }
         }
 
+        private void RefreshRouteTableLegacy()
+        {
+            var assemblies = AdditionalAssemblies == null ? new[] { AppAssembly } : new[] { AppAssembly }.Concat(AdditionalAssemblies);
```

Comment:
Does this need to compile against downlevel?  If not, Prepend will be cheaper than new[] + Concat here.